scalar: support cloning using the gvfs-helper

dscho opened the pull request:
This PR implements that crucial support in `scalar clone` where it allows to clone from Azure Repos (which do not support partial clone at time of writing).

To test this, we first have to hack up `test-gvfs-protocol`: The `test-gvfs-protocol` helper offers a full-fledged HTTP server that handles the GVFS protocol extensions, for use in the regression tests. However, it does not _actually_ serve Git repositories.

Therefore, this PR teaches it to do that, so that we can verify that our `contrib/scalar/` version of `scalar.exe` manages to clone from GVFS-enabled (and partial clone-disabled!) remotes in `t9050-scalar.sh`.

dscho committed:
test-gvfs-protocol: fix typos

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
WIP test-gvfs-protocol: also serve smart protocol

This comes in handy when we want to verify that `scalar clone` also
works against a GVFS-enabled remote repository.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
t5799: camelCase config settings

... because cSpell complained about it.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
t5799: spell 'web server' as two words

... because cSpell complained about it otherwise.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: verify that we can use a GVFS-enabled repository

Azure Repos does not support partial clones at the moment, but it does
support the GVFS protocol. To that end, the Microsoft fork of Git has a
`gvfs-helper` command that is optionally used to perform essentially the
same functionality as partial clone.

Let's verify that `scalar clone` detects that situation and enables the
GVFS helper.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! WIP: scalar: implement the `clone` subcommand

`scalar clone` actually clones into `<dir>/src/`, not into `<dir>`...

dscho committed:
fixup! scalar: test `scalar clone`

dscho committed:
WIP: scalar clone: support GVFS-enabled remote repositories

With this change, we come a big step closer to feature parity with
Scalar: this allows cloning from Azure Repos (which do not support
partial clones at time of writing).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: determine default branch name earlier

This prepares for a more stringent refspec with --single-branch.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: do not revert to protocol v1

There _was_ a bug in some Git version that prevented protocol v2 from
working nicely, but this bug has since been squashed.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar clone: respect --single-branch

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! WIP: scalar: implement the `clone` subcommand

We do need to clean up before leaving this function, even if there is an
error...

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! WIP: scalar: implement the `clone` subcommand

No need to let the Microsoft-specific `useGVFSHelper` config setting
bleed into the part of `contrib/scalar/` that we plan on upstreaming.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! WIP: scalar: implement the `clone` subcommand

Let's make sure that we check out the initial branch as tracking the
remote's default branch.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! TEST: make manual testing easier via the `test` command

This facilitates testing whether we can retrieve the information about
the cache server.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: add a test script

osx-clang fails with

	fatal: unable to access 'http://127.0.0.1:9050/': Failed to
	connect to 127.0.0.1 port 9050: Connection refused

Trying a different port (by virtue of using a different test
number) did not really help (see
https://github.com/microsoft/git/pull/325/checks?check_run_id=2379395998#step:5:425)
but it is a nicer number, anyway.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: verify that we can use a GVFS-enabled repository

When stopping the GVFS test daemon, let's show the server log in verbose
mode.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho commented:
/cc @jeffhostetler (with a lot of gratitude for helping me with this)

derrickstolee commented on contrib/scalar/scalar.c:
```diff
@@ -141,14 +141,56 @@ static int set_config(const char *file, const char *fmt, ...)
+	p = strstr(out.buf, needle);
+	if (p)
+		p += strlen(needle);
+	while (*p == '{') {
+		curly = strchrnul(p, '}');
```
Is there any concern about these stacking? ```{ { { { content } } } }```?

Seems like the response would be unexpected and your search for the global default string would only work with a _wider_ range of results, so I'm not too upset about this.

dscho replied:
Since the GVFS protocol design is essentially frozen, I think we will be able to rely on the `CacheServers` attribute being a shallow object. Read: I don't think we need to care about stacking.

derrickstolee replied:
Right. This works.

derrickstolee commented on contrib/scalar/scalar.c:
```diff
@@ -278,37 +322,57 @@ static int cmd_clone(int argc, const char **argv)
+	if (cache_server_url ||
+	    !retrieve_cache_server_url(dir, url, &cache_server_url)) {
+		if (set_config(config_path, "core.useGVFSHelper=true") ||
+		    set_config(config_path, "remote.origin.promisor") ||
+		    set_config(config_path,
+			       "remote.origin.partialCloneFilter") ||
```
I don't think these apply at the same time as `core.gvfs`.

derrickstolee replied:
Also, we should still use `core.useGVFSHelper=true` and `core.gvfs=150` when the `/gvfs/config` endpoint exists. We can use the origin instead of the cache server. This would show up with the Scalar functional tests because the test repo doesn't have a cache server.

This presents an interesting step of this effort: make the scalar functional tests consume this feature branch!

dscho replied:
I probably should clarify that this syntax (where there is no `=` in the value) _unsets_ the variable.

But I agree, it should not even be set in the first place. Will re-do it.

dscho replied:
I tried to address this via cca26c13cef. What do you think?

derrickstolee commented on contrib/scalar/t/t9099-scalar.sh:
```diff
@@ -0,0 +1,115 @@
+		cd using-gvfs/src &&
+		test_path_is_missing 1/2 &&
+		GIT_TRACE=$PWD/trace.txt git cat-file blob $second >actual &&
+		: verify that the gvfs-helper was invoked to fetch it &&
```
I've never seen this `:` as a comment before. Should we be more idiomatic and use `#`?

dscho replied:
I typically use `:` because that way, comments show up in the trace (which is helpful when analyzing CI breakages).

derrickstolee replied:
Interesting!

derrickstolee commented on t/helper/test-gvfs-protocol.c:
```diff
@@ -1357,7 +1357,7 @@ static enum worker_result req__read(struct req *req, int fd)
 
 	if (strcmp(http_version, "HTTP/1.1")) {
-		logerror("unsuported version '%s' (expecting HTTP/1.1)",
+		logerror("unsupported version '%s' (expecting HTTP/1.1)",
```
Good catch.

dscho replied:
cSpell for the win!

dscho committed:
fixup! scalar: add a test script

dscho committed:
fixup! scalar: verify that we can use a GVFS-enabled repository

dscho committed:
fixup! scalar clone: respect --single-branch

dscho committed:
scalar: avoid configuring GVFS helper preemptively

Instead of configuring it one way and then reverting it if we detect no
GVFS protocol support, first try to detect it, then use the right
settings right away.

Also: if no cache server was reported by the remote, we should configure
any, but we can _still_ use the GVFS endpoints.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar diagnose: include disk space information

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar unregister: stop FSMonitor daemon

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: more TODOs

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar test: verify `get_disk_info()`

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: add the `cache-server` command

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: refactor `is_unattended` into a helper function

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: use the same variable name for the cache directory as the .NET version

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
WIP: scalar clone: handle --local-cache-root default

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
WIP scalar clone: set ACLs on Windows

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar clone: add TODO that we need to register this with Scalar

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar unregister: stop FSMonitor daemon

We need to be careful: if the daemon was _already_ stopped, `--stop`
returns an error.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! WIP: scalar clone: handle --local-cache-root default

Need to be careful with `dir_inside_of()`...

dscho committed:
scalar: more TODOs

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
gvfs-helper: add the `endpoint` command

We already have the `config` command that accesses the `gvfs/config`
endpoint.

To implement `scalar`, we also need to be able to access the `vsts/info`
endpoint. Let's add a command to do precisely that.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: adjust TODOs

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: adjust TODOs

dscho committed:
scalar: implement a minimal JSON parser

No grown-up C project comes without their own JSON parser.

Just kidding!

We need to parse a JSON result when determining which cache server to
use. It would appear that searching for needles `"CacheServers":[`,
`"Url":"` and `"GlobalDefault":true` _happens_ to work right now, it is
fragile as it depends on no whitespace padding and on the order of the
fields remaining as-is.

Let's implement a super simple JSON parser (at the cost of being
slightly inefficient) for that purpose. To avoid allocating a ton of
memory, we implement a callback-based one. And to save on complexity,
let's not even bother validating the input properly (we will just go
ahead and instead rely on Azure Repos to produce correct JSON).

Note: An alternative would have been to use existing solutions such as
JSON-C, CentiJSON or JSMN. However, they are all a lot larger than the
current solution; The smallest, JSMN, which does not even provide parsed
string values (something we actually need) weighs in with 471 lines,
while we get away with 182 + 29 lines for the C and the header file,
respectively.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: replace fragile JSON parsing

When we access the `gvfs/config` endpoint, we need to actually parse the
JSON that comes back, lest we match something provided, say, in the
repository's description.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: do initialize `gvfs.sharedCache`

This finalizes the port of the `QueryVstsInfo()` function: we already
taught `gvfs-helper` to access the `vsts/info` endpoint on demand, we
implemented proper JSON parsing, and now it is time to hook it all up.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Add one more TODO

So I don't forget... ;-)

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: avoid error when registering scalar

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: the `cache-server` command needs a gitdir

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: document the `clone` subcommand

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: document `list`

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: document the `clone` subcommand

dscho committed:
squash! scalar: document `list`

scalar: document `list`, `register` and `unregister`

dscho committed:
scalar: document the remaining subcommands

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar: do initialize `gvfs.sharedCache`

dscho committed:
scalar: always use SHA-1 to generate the URL-based cache key

The principal idea of the cache key is that it should correspond to the
given URL, so that the same cache key is used no matter how often the
user clones the same repository.

However, if we change this depending what default hash algorithm is used
by Git, we will end up with different cache keys. Not what we want.

So let's fix this to SHA-1 for now (falling back to Git's default hash
algorithm if SHA-1 is not available, to future-proof this code).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: more TODOs

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: only try GVFS protocol on https:// URLs

Well, technically also the http:// protocol is allowed _when testing_...

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar test: fix on Windows

We need to use the "correct variant" of the working directory's path:
`git.exe` will understand only the Windows version of that path, while
`$PWD` will refer to the pseudo-Unix version of it.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar test: reduce side effects

With this change, --run=2,5,6 will work, otherwise 3 has to be included
(_just_ for the side effect of setting `second`).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
test-gvfs-protocol: prevent MSYS2 from mangling PATH_TRANSLATED

The value _does_ look like a Unix-style path, but no, MSYS2 must not be
allowed to convert that into a Windows path: `http-backend` needs it in
the unmodified form.

The MSYS2 runtime comes in when `git` is run via `bin-wrappers/git`
(which is a shell script).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! ci: also run `scalar` tests

Forgot the Windows side of things! And sure enough, they were broken
(but the past few commits fixed it).

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

derrickstolee commented on contrib/scalar/scalar.c:
```diff
@@ -913,7 +913,7 @@ static int add_or_remove_enlistment(int add)
 
 	return run_git(NULL, "config", "--global",
 		       add ? "--add" : "--unset",
-		       "--fixed-value", "scalar.repo",
+		       "scalar.repo",
```
Really dropping `--fixed-value`? This option was created exactly for cases like this.

dscho replied:
```
12:16:01.574256 git.c:447               trace: built-in: git config --global --add --fixed-value scalar.repo C:/Users/me/repos/ForTests-123/src
error: --fixed-value only applies with 'value-pattern'
[...]
```

So yes, really dropping `--fixed-value` here. I keep it for the query, but not for setting it (because I want the command to succeed 😁).

derrickstolee replied:
We need it for the `--unset` option. Perhaps this should be split with an `if (add)`.

dscho replied:
Oh, I see, we need it for the `--unset` case. But `--fixed-value` is incompatible with the `--add` case. So I replaced `"--fixed-value"` with `add ? "--no-fixed-value" : "--fixed-value"`.

dscho committed:
scalar: remove unused code

I thought that I'd have to call `scalar.exe` in a child process, but it
turns out that that's not actually necessary.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: generate usage string dynamically

The subcommand list was somewhat stale; With this change, it cannot go
stale anymore.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fsmonitor: only enable it in non-bare repositories

The entire point of the FSMonitor is to monitor the worktree changes in
a more efficient manner than `lstat()`ing all worktree files every time
we refresh the index.

But if there is no worktree, FSMonitor has nothing to monitor.

So let's ignore if an FSMonitor is configured (e.g. in `~/.gitconfig`)
and we're running in a repository without worktree.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
Remove TODO: FSMonitor is now disabled in bare repositories

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar diagnose: write into the correct location again

When I changed the code to initialize the worktree in the `src/`
subdirectory, I forgot to adjust the directory to which the diagnose
output is written...

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar diagnose: report the absolute path

When running `scalar diagnose` in a directory other than the top-level
worktree directory, seeing `../src/.scalarDiagnose/[...]` can be
misleadig.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar diagnose: show a spinner while staging content

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar diagnose: report the absolute path

dscho committed:
scalar: prepare for registering just-cloned path

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar clone: register the just-cloned worktree

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar unregister: only stop built-in FSMonitor if there is one

On unsupported platforms, it does not make sense to stop the FSMonitor.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! scalar diagnose: show a spinner while staging content

We're supposed to check the return value of `write()` and act on errors
;-)

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
scalar: accept -C and -c options before the subcommand

The `git` executable has these two very useful options:

-C <directory>:
	switch to the specified directory before performing any actions

-c <key>=<value>:
	temporarily configure this setting for the duration of the
	specified scalar subcommand

With this commit, we teach the `scalar` executable the same trick.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

dscho committed:
fixup! ci: also run `scalar` tests

Need to include `scalar` in the artifacts copied to the test jobs ;-)

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>

derrickstolee commented:
I noticed in testing that `scalar clone` reports usage as follows:

```ShellSession
$ scalar clone
fatal: need a URL

usage: git clone [<options>] [--] <repo> [<dir>]

    --cache-server-url <url>
                          the url or friendly name of the cache server
    -b, --branch <branch>
                          branch to checkout after clone
    --single-branch       only download metadata for the branch that will be checked out
    --no-fetch-commits-and-trees
                          skip fetching commits and trees after clone
    --local-cache-path <path>
                          override the path for the local Scalar cache
    --full-clone          when cloning, create full working directory
```

The important part is that the usage says `git clone` instead of `scalar clone`.

_Edit:_ The issue appears to be at [this line in `scalar.c`](https://github.com/microsoft/git/blob/70627d31e2827a344a3f6fc53826149e4a4a91ec/contrib/scalar/scalar.c#L490).

derrickstolee commented on contrib/scalar/scalar.c:
```diff
@@ -233,7 +502,7 @@ static int cmd_clone(int argc, const char **argv)
 	if (argc == 2) {
 		url = argv[0];
-		dir = xstrdup(argv[1]);
+		root = xstrdup(argv[1]);
```
(Actually for the context above)

I think what you want is this:

```
	argc = parse_options(argc, argv, NULL, clone_options, clone_usage,
			     PARSE_OPT_KEEP_UNKNOWN);
```

Specifically, the Scalar functional tests present a different order:

`scalar clone <url> <dir> <other-options>`

derrickstolee replied:
After swapping the order, however, I get the following error!!!

```

C:\_git\scalar\src>scalar clone --branch "FunctionalTests/20180214" --local-cache-path "C:\Repos\ScalarFunctionalTests\enlistment\..\.scalarCache"  "https://gvfs.visualstudio.com/ci/_git/ForTests" "C:\Repos\ScalarFunctionalTests\enlistment\5950128c27f94410b75b"
Initialized empty Git repository in C:/Repos/ScalarFunctionalTests/enlistment/5950128c27f94410b75b/src/.git/
Receiving vsts/info (bytes received):   0% (0/89520), done.
Receiving vsts/info (bytes received): 100% (776/776), done.
Receiving gvfs/config (bytes received): 113, done.
Prefetch 0 (1970-01-01 00:00:00 +0000) (bytes received): 228997, done.
Receiving packfile 1/1 with 7 objects (bytes received): 8473, done.
Already on 'FunctionalTests/20180214'
Branch 'FunctionalTests/20180214' set up to track remote branch 'FunctionalTests/20180214' from 'origin'.
Your branch is up to date with 'origin/FunctionalTests/20180214'.
Attempt to free already freed memory block 0000012354A00B28 - aborting!
```

This issue only happend in `cmd` not in Git Bash, strangely.

dscho replied:
Oh... Good point. I shall change that, and adjust the regression test so that this is exercised.

dscho replied:
I actually only had to drop the `PARSE_OPT_STOP_AT_NON_OPTION`, and not include `PARSE_OPT_KEEP_UNKNOWN`.

dscho commented:
> _Edit:_ The issue appears to be at [this line in `scalar.c`](https://github.com/microsoft/git/blob/70627d31e2827a344a3f6fc53826149e4a4a91ec/contrib/scalar/scalar.c#L490).

D'OH! Thanks for being so diligent!

dscho committed:
fixup! scalar: add the `cache-server` command

It's the `scalar` command, not the `git` command. Thanks, Stolee, for
pointing this out to me!

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

dscho committed:
fixup! scalar: add the `cache-server` command

Scalar is traditionally not strict about requiring the options to come
before the non-option parameters.

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

dscho committed:
fixup! WIP: scalar: implement the `clone` subcommand

It's the `scalar` command, not the `git` command. Thanks, Stolee, for
pointing this out to me!

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

dscho committed:
fixup! WIP: scalar: implement the `clone` subcommand

Scalar is traditionally not strict about requiring the options to come
before the non-option parameters.

Also, we must not keep the double-dash, as we're not handling it.

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

dscho committed:
scalar test: verify that options can come after parameters

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

dscho committed:
scalar test: verify that the dash-dash is handled gracefully

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>

dscho committed:
fixup! ci: also run `scalar` tests

Also need to include `scalar` in the artifacts copied to the vs-test
jobs ;-)

Signed-off-by: Johannes Schindelin <johasc@microsoft.com>
